data_sources/amanda_closure_publishing.py

```diff
@@ -0,0 +1,234 @@
+    }
+
+    return feed_info
+
+
+def main():
+    # Getting AMANDA data
+    data = get_amanda_data(turp_query)
+    df = pd.DataFrame(data)
+    logger.info(f"Downloaded {len(df['FOLDERRSN'].unique())} TURP permits")
+    data = get_amanda_data(excavation_permits)
+    df = pd.concat([df, pd.DataFrame(data)])
+    logger.info(f"Downloaded {len(df['FOLDERRSN'].unique())} EX permits")
+
+    df = df.apply(get_start_end_date, axis=1)
+    segments = df[
+        df["CLOSURE_TYPE"].isin(
+            ["Closure : Full Road", "Traffic Lane : Dimensions", "Open Cuts : Street"]
+        )
+    ]["SEGMENT_ID"].unique()
+    segment_info = get_geometry(segments)
+    segment_lookup = {}
+
+    # Generating a dictionary of street segment IDs
+    for s in segment_info:
+        segment_lookup[int(s["segment_id"])] = s
+
+    # Generating UUIDs data sources
+    amanda_turp_id = str(uuid.uuid5(uuid.NAMESPACE_OID, "COA_AMANDA_TURP"))
+    amanda_ex_id = str(uuid.uuid5(uuid.NAMESPACE_OID, "COA_AMANDA_EX"))
+
+    central_time_zone = pytz.timezone("US/Central")
+    current_time = datetime.datetime.now(central_time_zone)
+    df["START_DATE"] = pd.to_datetime(df["START_DATE"]).dt.tz_localize(
+        central_time_zone
+    )
+    df["END_DATE"] = pd.to_datetime(df["END_DATE"]).dt.tz_localize(central_time_zone)
+
+    work_zones = []
+    permits = df["FOLDERRSN"].unique()
+    for p in permits:
+        closures = df[df["FOLDERRSN"] == p]
+        permit_type = closures["FOLDERTYPE"].iloc[0]
+        folderdesc = closures["FOLDERDESCRIPTION"].iloc[0]
+        foldername = closures["FOLDERNAME"].iloc[0]
+        subtype = closures["SUBCODE"].iloc[0]
+        workcode = closures["WORKCODE"].iloc[0]
+        if permit_type == "RW":
+            # Filtering out details from franchise utilities.
+            if subtype == 50500 and workcode in (50570, 50575, 50580):
+                description = f"Temporary use of Right of Way Permit has been issued for this location."
+                name = "WorkZone Event"
+            else:
+                description = f"Temporary use of Right of Way Permit has been issued for this location. \n Details: {folderdesc}"
+                name = foldername
+            data_source_id = amanda_turp_id
+        elif permit_type == "EX":
+            if subtype == 50685:
+                description = f"Excavation Permit has been issued for this location."
+                name = "WorkZone Event"
+            else:
+                description = f"Excavation Permit has been issued for this location. \n Details: {folderdesc}"
+                name = foldername
+
+            data_source_id = amanda_ex_id
+
+        segments = closures["SEGMENT_ID"].unique()
+
+        start_date = closures["START_DATE"].iloc[0]
+        end_date = closures["END_DATE"].iloc[0]
+
+        # checking if the closure is some time in the future.
+        # adding one hour to the end time to help inform consumers that the work zone has officially ended.
```

Comment:
is the end time more like a targeted end time, and this 1 hour buffer is to hopefully ensure they are fully done?

Reply:
oh I see this is going to run hourly, got it

Reply:
This is what the spec suggests we do (even though all our permits end at midnight)

> Feed producers should consider including a completed road event with a verified start and end date in a public feed for a at least an hour or one feed refresh cycle (whichever is longer) to help inform consumers that the work zone has officially ended.

https://github.com/usdot-jpo-ode/wzdx/blob/main/Creating_a_WZDx_Feed.md#implementation-guidance